### PR TITLE
WIP: Env scaler 

### PR DESCRIPTION
**Pull request by nehayward**
**What does this pull request do?**
Adds scaler functionality to env.

**How should this be tested?**

**Checklist**
- [ ] Unit tests
- [ ] Smoke tests (if applicable)
- [ ] System tests (if applicable)
- [ ] Documentation (if applicable)

closes #
(reference the issue(s) this pull request closes)

### CODE REVIEW

File: api/provider/aws/common.go [hunk 1]
```diff
@@ -243,3 +243,47 @@ func listClusterTaskARNs(ecsapi ecsiface.ECSAPI, clusterName, startedBy, status
 
 	return taskARNs, nil
 }
+
+func listContainerInstances(ecsapi ecsiface.ECSAPI, clusterName string) ([]string, error) {
```
Comment (zpatrick) [inline api/provider/aws/common.go]: why is this is `common.go` - won't this only be used by the scaler?

File: api/provider/aws/common.go [hunk 2]
```diff
@@ -243,3 +243,47 @@ func listClusterTaskARNs(ecsapi ecsiface.ECSAPI, clusterName, startedBy, status
...
+	containerInstanceARNs := []string{}
+	fn := func(output *ecs.ListContainerInstancesOutput, lastPage bool) bool {
+		for _, c := range output.ContainerInstanceArns {
```
Comment (zpatrick) [inline api/provider/aws/common.go]: Do we need to filter by the state of the container instances? For example, will this list container instances that aren't actively running?

File: api/provider/aws/common.go [hunk 3]
```diff
@@ -243,3 +243,47 @@ func listClusterTaskARNs(ecsapi ecsiface.ECSAPI, clusterName, startedBy, status
...
+	input.SetCluster(clusterName)
+
+	if err := ecsapi.ListContainerInstancesPages(input, fn); err != nil {
+		return nil, err
+	}
+
+	return containerInstanceARNs, nil
+}
+
+func listClusterServiceNames(ecsapi ecsiface.ECSAPI, clusterNames []string) ([]string, error) {
```
Comment (zpatrick) [inline api/provider/aws/common.go]: don't we already have a helper that does this?

File: api/provider/aws/common.go (hunk 1 above)
Comment (nehayward) [reply api/provider/aws/common.go]: @sesh-kebab

File: api/provider/aws/environment_scaler.go [hunk 4]
```diff
@@ -1,14 +1,629 @@
...
+				}
+			}
+		}
+
+		inUse := aws.Int64Value(instance.PendingTasksCount)+aws.Int64Value(instance.RunningTasksCount) > 0
+		r := &models.ResourceProvider{
+			ID:              aws.StringValue(instance.Ec2InstanceId),
+			InUse:           inUse,
+			UsedPorts:       usedPorts,
+			AvailableMemory: fmt.Sprintf("%v", availableMemory), //todo: this is definitely not correct
```
Comment (jparsons04) [inline api/provider/aws/environment_scaler.go]: Assuming that the `AvailableMemory` field here accepts a string that is a numeric value and the unit (for instance, 10MB), could you not leverage the [Format()](https://github.com/zpatrick/go-bytesize/blob/master/bytesize.go#L112) method in go-bytesize?

File: api/provider/aws/environment_scaler.go [hunk 5]
```diff
@@ -1,14 +1,629 @@
...
+		containerInstanceARNs = append(containerInstanceARNs, output.ContainerInstanceArns...)
+
+		return !lastPage
+	}
+
+	if err := e.Client.ECS.ListContainerInstancesPages(listContainerInstancesInput, listContainerInstancesPagesFN); err != nil {
+		return nil, err
+	}
+
+	describeContainerInstancesInput := &ecs.DescribeContainerInstancesInput{}
```
Comment (sesh-kebab) [inline api/provider/aws/environment_scaler.go]: Might as well check for a 0 length of `containerInstanceARNs` returned. If 0 we don't need to proceed any further. I.e., 

```go
result := []*models.ResourceProvider{}
if len(containerInstanceARNs) == 0 {
	return result, nil
}
```

File: api/provider/aws/environment_scaler.go [hunk 6]
```diff
@@ -1,14 +1,629 @@
...
-	return fmt.Errorf("EnvironmentScaler not implemented")
+	// GET RESOURCE PROVIDERS
+	clusterName := addLayer0Prefix(e.Config.Instance(), environmentID)
+
+	resourceProviders, err := e.getResourceProviders(clusterName)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] resourceProviders for env '%s': %#v", environmentID, resourceProviders)
```
Comment (jparsons04) [inline api/provider/aws/environment_scaler.go]: [nit] Could we print this only if some sort of debug flag is set in configuration?

File: api/provider/aws/environment_scaler.go [hunk 7]
```diff
@@ -1,14 +1,629 @@
...
+
+	// GET RESOURCE CONSUMERS
+	resourceConsumers, err := e.getResourceConsumers(clusterName)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] resourceConsumers for env '%s': %#v", environmentID, resourceConsumers)
```
Comment (jparsons04) [inline api/provider/aws/environment_scaler.go]: [nit] Same question about debug flags

File: api/provider/aws/environment_scaler.go [hunk 8]
```diff
@@ -1,14 +1,629 @@
...
+
+	info, err := e.scale(clusterName, resourceProviders, resourceConsumers)
+	if err != nil {
+		return err
+	}
+	fmt.Println(info)
```
Comment (jparsons04) [inline api/provider/aws/environment_scaler.go]: Do we intend to output this model to the terminal long term?

File: api/provider/aws/environment_scaler.go (hunk 6 above)
Comment (jparsons04) [reply api/provider/aws/environment_scaler.go]: Disregard. I can see in several other areas within this package that DEBUG statements are freely printed. It must just be a matter of log level specification for the event happening. Seeing this, I don't think the debug flag question is worth looking into at this point.

File: api/provider/aws/environment_scaler.go (hunk 7 above)
Comment (jparsons04) [reply api/provider/aws/environment_scaler.go]: Disregard for the same reason as stated above.

File: api/provider/aws/environment_scaler.go (hunk 6 above)
Comment (zpatrick) [reply api/provider/aws/environment_scaler.go]: statements that start with `[DEBUG]` only print if the debug log level is set

File: api/provider/aws/environment_scaler.go [hunk 9]
```diff
@@ -1,14 +1,629 @@
...
+	// check if we need to scale up
+	for _, consumer := range consumers {
+		hasRoom := false
+
+		// first, sort by memory so we pack tasks by memory as tightly as possible
+		sortProvidersByMemory(providers)
+
+		// next, place any unused providers in the back of the list
+		// that way, we can can delete them if we avoid placing any tasks in them
+		sortProvidersByUsage(providers)
```
Comment (jparsons04) [inline api/provider/aws/environment_scaler.go]: I may be misunderstanding how the sort package works with this, so please correct me if I'm wrong. If you sort the same object (the length and members are unchanged after the first sort) two consecutive times, mathematically speaking, wouldn't the second sort just simply override what the first sort did?  These functions don't appear to manipulate (add or remove) any items into the slice that I can see, just simply sorting them. Assuming the length and members of `providers` is unchanged after the first sort, I don't think the second sort operation cares what the first sort did.

Comment (jparsons04) [reply api/provider/aws/environment_scaler.go]: I'm implying here that only `sortProvidersByUsage` is required.

File: api/provider/aws/environment_scaler.go [hunk 10]
```diff
@@ -1,14 +1,629 @@
...
+	default:
+		log.Printf("Environment %s is at desired scale of %d. No scaling action required.", clusterName, scale)
+		return currentCapacity, nil
+	}
+}
+
+func (e *EnvironmentScaler) scaleUp(ecsEnvironmentID string, scale int, asg *autoscaling.Group) (int, error) {
+	maxCapacity := int(aws.Int64Value(asg.MaxSize))
+	if scale > maxCapacity {
+		input := &autoscaling.UpdateAutoScalingGroupInput{}
```
Comment (jparsons04) [inline api/provider/aws/environment_scaler.go]: [nit] For consistency, since you have `log.Printf()` statements for `scaleDown` when scale < minCapacity, I think similar logging statements should be present in this method when scale > maxCapacity.

File: api/provider/aws/test_aws/environment_scaler_test.go [hunk 11]
```diff
@@ -0,0 +1,9 @@
+package test_aws
+
+import (
+	"testing"
+)
+
+func TestEnvironmentScale(t *testing.T) {
```
Comment (zpatrick) [inline api/provider/aws/test_aws/environment_scaler_test.go]: You should add in the old tests back in verbatim - they are really extensive and do a good job of catching corner cases

File: api/provider/aws/environment_scaler.go (hunk 9 above)
Comment (zpatrick) [reply api/provider/aws/environment_scaler.go]: @jparsons04 `sortProvidersByMemory` uses an integer comparison, `sortProvidersByUsage` just does a boolean comparison. Doing these sorts allows unused providers to bubble to the end of the list - e.g. it's the last candidate we should try to place a resource in. 

This is important because if we have unused providers at the end of this process, we know we can scale down.

File: api/provider/aws/common.go (hunk 1 above)
Comment (sesh-kebab) [reply api/provider/aws/common.go]: Changes to common can be removed. It doesn't look like the latest set of changes that @nehayward and @tlake worked on use these methods.

File: api/provider/aws/common.go (hunk 2 above)
Comment (sesh-kebab) [reply api/provider/aws/common.go]: Agreed - although this method isn't being used any longer, the same request being made from `environment_scaler.go` could would want to specify the `Status` property of the `ListContainerInstancesInput` (where status == ACTIVE)

File: api/provider/aws/common.go (hunk 3 above)
Comment (sesh-kebab) [reply api/provider/aws/common.go]: Not that i'm aware of. That was I think one of the biggest issues with the existing code where it looked like we were getting services from all environments and then filter for a particular environment. I'm sure this would have been a major if not the biggest contributor to the rate limits we were seeing.

Comment (sesh-kebab) [reply api/provider/aws/common.go]: Also same thing applies - this code is no longer used.

File: api/provider/aws/environment_scaler.go [hunk 12]
```diff
@@ -1,14 +1,629 @@
...
+	"github.com/quintilesims/layer0/common/errors"
+	"github.com/quintilesims/layer0/common/models"
+)
 
 type EnvironmentScaler struct {
+	Client              *awsc.Client
+	EnvironmentProvider provider.EnvironmentProvider
+	ServiceProvider     provider.ServiceProvider
+	TaskProvider        provider.TaskProvider
+	JobStore            job.Store
```
Comment (sesh-kebab) [inline api/provider/aws/environment_scaler.go]: We shouldn't need the job store.

File: api/provider/aws/environment_scaler.go [hunk 13]
```diff
@@ -1,14 +1,629 @@
...
+	Config              config.APIConfig
+	deployCache         map[string][]models.ResourceConsumer
```
Comment (sesh-kebab) [inline api/provider/aws/environment_scaler.go]: Not thread safe (or should that be go routine safe). Not that scale can be called concurrently via the JobRunner and the Scaler timer. Maybe a RWMutex or a concurrent collection?

File: api/provider/aws/environment_scaler.go [hunk 14]
```diff
@@ -1,14 +1,629 @@
...
+			AvailableMemory: provider.AvailableMemory,
+		}
+		providerModels[i] = newProvider
+	}
+
+	return providerModels
+}
+
+func (e *EnvironmentScaler) getResourceProviders(clusterName string) ([]*models.ResourceProvider, error) {
+	listContainerInstancesInput := &ecs.ListContainerInstancesInput{}
```
Comment (sesh-kebab) [inline api/provider/aws/environment_scaler.go]: [nit] I know we have a guideline of not calling helper methods from other helper methods but it sure would be cleaner in my opinion to wrap the aws calls in their own method. So getResourceProviders would look something like

```go
func (e *EnvironmentScaler) getResourceProviders(clusterName string) ([]*models.ResourceProvider, error) {
   clusterInstances, err := getContainerInstances(clusterName)
   if err != nil ...

```

File: api/provider/aws/environment_scaler.go [hunk 15]
```diff
@@ -1,14 +1,629 @@
...
+		return nil, err
+	}
+
+	result := []*models.ResourceProvider{}
+	if len(output.ContainerInstances) == 0 {
+		return result, nil
+	}
+
+	for _, instance := range output.ContainerInstances {
+		if !aws.BoolValue(instance.AgentConnected) {
```
Comment (sesh-kebab) [inline api/provider/aws/environment_scaler.go]: As discussed in slack - there is an incorrect assumption being made here. Either remove this (as we don't handle this case when container instance count == min cluster count. Or would need to check the connection state for a duration. I.e., if disconnected for more than 3 minutes then consider it permanently disconnected.

In practice - checking for a certain duration will add considerable complexity which I don't think is warranted given:

1. we don't do anything even if the agent is disconnected when current cluster count == min cluster count
2. how often does this situation actually happen?

File: api/provider/aws/environment_scaler.go [hunk 16]
```diff
@@ -1,14 +1,629 @@
...
+			continue
+		}
+
+		if aws.StringValue(instance.Status) != "ACTIVE" {
```
Comment (sesh-kebab) [inline api/provider/aws/environment_scaler.go]: Can be added to the request rather than filtering for it here.

File: api/provider/aws/environment_scaler.go [hunk 17]
```diff
@@ -1,14 +1,629 @@
...
+		return !lastPage
+	}
+
+	if err := e.Client.ECS.ListServicesPages(listServicesInput, listServicesPagesFN); err != nil {
+		return nil, err
+	}
+
+	services := []*ecs.Service{}
+	if len(serviceARNs) > 0 {
+		// The SDK states that you can specify up to 10 services in one DescribeServices operation:
```
Comment (sesh-kebab) [inline api/provider/aws/environment_scaler.go]: 👍

File: api/provider/aws/environment_scaler.go [hunk 18]
```diff
@@ -1,14 +1,629 @@
...
+			if err != nil {
+				return nil, err
+			}
+
+			resourceConsumers = append(resourceConsumers, c...)
+		}
+	}
+
+	// GET PENDING TASK RESOURCE CONSUMERS IN ECS
+	taskSummaries, err := e.TaskProvider.List()
```
Comment (sesh-kebab) [inline api/provider/aws/environment_scaler.go]: ❗️ This is super inefficient. Services had a similar pattern (in develop) where all entities are querie, cluster by cluster for all the clusters. Then you are throwing away most of the results by filtering for a single environment cluster.

Lets only query the tasks from the environment that we're currently scaling.

File: api/provider/aws/environment_scaler.go [hunk 19]
```diff
@@ -1,14 +1,629 @@
...
+			if err != nil {
+				return nil, err
+			}
+
+			resourceConsumers = append(resourceConsumers, taskResourceConsumers...)
+		}
+	}
+
+	// GET PENDING TASK RESOURCE CONSUMERS IN JOBS
+	jobs, err := e.JobStore.SelectAll()
```
Comment (sesh-kebab) [inline api/provider/aws/environment_scaler.go]: As discussed in slack - we don't need to consider jobs as resource consumers any longer as we execute the jobs via the JobRunner and no ecs tasks.

File: api/provider/aws/environment_scaler.go [hunk 20]
```diff
@@ -1,14 +1,629 @@
...
+		return nil, err
+	}
+
+	var instanceType string
+	if len(config.LaunchConfigurations) > 0 {
+		instanceType = aws.StringValue(config.LaunchConfigurations[0].InstanceType)
+	}
+
+	// these ports are automatically used by the ecs agent
+	defaultPorts := []int{
```
Comment (sesh-kebab) [inline api/provider/aws/environment_scaler.go]: Can we move this to a const at the top of the file.

File: api/provider/aws/environment_scaler.go [hunk 21]
```diff
@@ -1,14 +1,629 @@
...
+				}
+			}
+		}
+	}
+
+	return resourceConsumers, nil
+}
+
+func (e *EnvironmentScaler) calculateNewProvider(clusterName string) (*models.ResourceProvider, error) {
+	input := &autoscaling.CreateAutoScalingGroupInput{}
```
Comment (sesh-kebab) [inline api/provider/aws/environment_scaler.go]: Why are we creating an AutoScalingGroup here? The whole purpose seems to be to get the InstanceType from a launch config. Can we not get this from environment.size?

1. Create ASG
2. Describe the launch config for the newly created group

File: api/provider/aws/environment_scaler.go [hunk 22]
```diff
@@ -1,14 +1,629 @@
...
+	resource.ID = "<new instance>"
+	resource.InUse = false
+	resource.UsedPorts = defaultPorts
+	resource.AvailableMemory = instanceType
+	return resource, nil
+}
+
+func (e *EnvironmentScaler) getContainerResourceFromDeploy(deployID string) ([]models.ResourceConsumer, error) {
+	// use some kind of deploy cache
+	if consumers, ok := e.deployCache[deployID]; ok {
```
Comment (sesh-kebab) [inline api/provider/aws/environment_scaler.go]: See previous comment about thread-safety. The easiest approach I suppose could be to have two instances of scaler (one for scaling and one for the job runner).

File: api/provider/aws/environment_scaler.go (hunk 9 above)
Comment (jparsons04) [reply api/provider/aws/environment_scaler.go]: Ok, then that makes sense if the sorts are operating on numeric then boolean comparisons.

File: api/provider/aws/environment_scaler.go (hunk 4 above)
Comment (nehayward) [reply api/provider/aws/environment_scaler.go]: Good catch, It looks like we could.

File: api/provider/aws/test_aws/environment_scaler_test.go (hunk 11 above)
Comment (nehayward) [reply api/provider/aws/test_aws/environment_scaler_test.go]: Thanks, I was wondering about that. I'll start working on that too.

File: api/provider/aws/environment_scaler.go (hunk 18 above)
Comment (nehayward) [reply api/provider/aws/environment_scaler.go]: Would the aws api be used directly to filter input for single env

File: api/provider/aws/environment_scaler.go (hunk 21 above)
Comment (nehayward) [reply api/provider/aws/environment_scaler.go]: We could remove #L378-399 and then just get available memory from env size, using tag store right?

File: api/provider/aws/environment_scaler.go (hunk 18 above)
Comment (sesh-kebab) [reply api/provider/aws/environment_scaler.go]: Yup exactly that. If we look in the implementation for `e.TaskProvivider.List()`, it does an equivalent of an `environment.List()` and then calls a helper method called `listClusterTaskARNs` for each cluster/env. I think either exposing the listClusterTaskARNs to the provider somehow or duplicate the code in the scaler would be good approach (probably the latter might be simple to start with).

File: api/provider/aws/environment_scaler.go (hunk 13 above)
Comment (nehayward) [reply api/provider/aws/environment_scaler.go]: Hmm maybe use this package? https://github.com/orcaman/concurrent-map

Comment (zpatrick) [reply api/provider/aws/environment_scaler.go]: or possible [go-cache](https://github.com/zpatrick/go-cache) :trollface:

File: api/provider/aws/environment_scaler.go [hunk 23]
```diff
@@ -358,6 +362,33 @@ func (e *EnvironmentScaler) getResourceConsumers(clusterName string) ([]models.R
...
 	}
 
+	// GET PENDING TASK RESOURCE CONSUMERS IN JOBS
+	jobs, err := e.JobStore.SelectAll()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, job := range jobs {
+		if job.Type == models.CreateDeployJob {
```
Comment (zpatrick) [inline api/provider/aws/environment_scaler.go]: `CreateTaskJob`

File: api/provider/aws/environment_scaler.go (hunk 19 above)
Comment (zpatrick) [reply api/provider/aws/environment_scaler.go]: Not all jobs - but the `CreateTaskJobs` that are `InProgress` do have pending resources

File: api/provider/aws/environment_scaler.go [hunk 24]
```diff
@@ -1,14 +1,656 @@
...
+
+	_, err = e.scale(clusterName, resourceProviders, resourceConsumers)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *EnvironmentScaler) scale(clusterName string, providers []*models.Resource, consumers []*models.Resource) (*models.ScalerRunInfo, error) {
```
Comment (diemonster) [inline api/provider/aws/environment_scaler.go]: It would be nice to reduce the overall spaghetti code and/or improve the naming of some of these functions. I still really like the pattern we landed on in the `provider/aws` layer where we used the main exported function (in this case, `Scale`) as a 'master function' for calling all the related helpers. We could do something similar here and I think it would greatly aid clarity.

File: api/provider/aws/environment_scaler.go [hunk 25]
```diff
@@ -1,14 +1,656 @@
...
+		ScaleBeforeRun:          scaleBeforeRun,
+		DesiredScaleAfterRun:    desiredScale,
+		ActualScaleAfterRun:     actualScale,
+		UnusedResourceProviders: len(unusedProviders),
+	}
+
+	return info, errors.MultiError(errs)
+}
+
+func resourceModels(resources []*models.Resource) []models.Resource {
```
Comment (diemonster) [inline api/provider/aws/environment_scaler.go]: This function needs a better name

File: api/provider/aws/environment_scaler.go [hunk 26]
```diff
@@ -1,14 +1,656 @@
...
+			CPU:    availableCPU.Megabytes(),
+		}
+
+		result = append(result, r)
+	}
+
+	return result, nil
+}
+
+func (e *EnvironmentScaler) getResourceConsumers(clusterName string) ([]*models.Resource, error) {
```
Comment (diemonster) [inline api/provider/aws/environment_scaler.go]: This function needs to be broken up

File: api/provider/aws/test_aws/environment_scaler_test.go [hunk 27]
```diff
@@ -0,0 +1,10 @@
+package test_aws
+
+import (
+	"testing"
+)
+
+// Add Old Test Verbatim from Dev PR
```
Comment (diemonster) [inline api/provider/aws/test_aws/environment_scaler_test.go]: Thinking those tests will need to be totally rewritten, I don't know how worthwhile they were to begin with.

File: api/provider/aws/environment_scaler.go [hunk 28]
```diff
@@ -1,14 +1,656 @@
...
+	}
+
+	if cpuInstanceSize > memoryInstanceSize {
+		return CPU
+	}
+
+	return Memory
+}
+
+func hasResourcesFor(consumer models.Resource, provider models.Resource) bool {
```
Comment (diemonster) [inline api/provider/aws/environment_scaler.go]: needs a better function name

File: api/provider/aws/environment_scaler.go [hunk 29]
```diff
@@ -1,14 +1,656 @@
...
+			if wanted == used {
+				return false
+			}
+		}
+	}
+
+	return consumer.Memory <= provider.Memory
+}
+
+func subtractResourcesFor(consumer *models.Resource, provider *models.Resource) error {
```
Comment (diemonster) [inline api/provider/aws/environment_scaler.go]: needs a better function name

File: api/provider/aws/environment_scaler.go [hunk 30]
```diff
@@ -1,14 +1,656 @@
...
+
+	provider.Ports = append(provider.Ports, consumer.Ports...)
+	provider.Memory -= consumer.Memory
+	provider.CPU -= consumer.CPU
+	provider.InUse = true
+
+	return nil
+}
+
+func sortByMemory(resources ...[]*models.Resource) {
```
Comment (diemonster) [inline api/provider/aws/environment_scaler.go]: More naming gripes. With all of these sort functions: sort how?

File: api/provider/aws/test_aws/environment_scaler_test.go (hunk 27 above)
Comment (nehayward) [reply api/provider/aws/test_aws/environment_scaler_test.go]: I’ve already started on testing and I’m using all the mockaws calls. I’ll put something up today to view. However I’m waiting on tlake to finish scaler so it’ll have to be adjusted. Also breaking up the code will help.

File: api/provider/aws/environment_scaler.go (hunk 24 above)
Comment (nehayward) [reply api/provider/aws/environment_scaler.go]: I agree completely.

Comment (nehayward) [reply api/provider/aws/environment_scaler.go]: I agree completely.

File: api/provider/aws/test_aws/environment_scaler_test.go [hunk 31]
```diff
@@ -0,0 +1,351 @@
...
+	ResourceConsumers []models.Resource
+}
+
+// Add Old Test Verbatim from Dev PR
+
+func (e *EnvironmentScalerUnitTest) Run(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockAWS := awsc.NewMockClient(ctrl)
```
Comment (zpatrick) [inline api/provider/aws/test_aws/environment_scaler_test.go]: this all needs to be split out - we should be testing the scaling algorithm as a single unit

File: api/provider/aws/test_aws/environment_scaler_test.go [hunk 32]
```diff
@@ -0,0 +1,351 @@
...
+	// 	Return(nil)
+
+	environmentScaler := provider.NewEnvironmentScaler(mockAWS.Client(), mockEnvironmentProvider, mockServiceProvider, mockTaskProvider, mockJobStore, mockConfig)
+
+	if err := environmentScaler.Scale("env_id"); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestResourceManagerScaleUp_noProviders(t *testing.T) {
```
Comment (zpatrick) [inline api/provider/aws/test_aws/environment_scaler_test.go]: Seems like these should jsut be table tests

File: api/provider/aws/test_aws/environment_scaler_test.go [hunk 33]
```diff
@@ -0,0 +1,351 @@
...
+				InUse:  true}},
+		ResourceConsumers: []models.Resource{
+			{Ports: []int{80}},
+		},
+	}
+
+	test.Run(t)
+}
+
+func TestResourceManagerNoScale_noResourceConsumers(t *testing.T) {
```
Comment (zpatrick) [inline api/provider/aws/test_aws/environment_scaler_test.go]: Seems like we are missing a lot of test cases

File: common/models/resource.go [hunk 34]
```diff
@@ -0,0 +1,9 @@
+package models
+
+type Resource struct {
```
Comment (zpatrick) [inline common/models/resource.go]: Not sure why this is a model - we don't use these objects to display anything to users

File: common/models/scaler_run_info.go [hunk 35]
```diff
@@ -1,11 +1,11 @@
 package models
 
 type ScalerRunInfo struct {
```
Comment (zpatrick) [inline common/models/scaler_run_info.go]: [nit] I thought we were removing this?

File: common/models/resource.go (hunk 34 above)
Comment (diemonster) [reply common/models/resource.go]: Good point. Maybe it's a holdover from when we talked (?) about having the scaler be a standalone thing? Regardless, I agree, these structs should be moved locally.

File: api/provider/aws/environment_scaler.go (hunk 28 above)
Comment (zpatrick) [reply api/provider/aws/environment_scaler.go]: these functions already exist on the resource consumer - why are they duplicated here?

File: api/provider/aws/environment_scaler.go (hunk 29 above)
Comment (zpatrick) [reply api/provider/aws/environment_scaler.go]: ditto

File: api/provider/aws/environment_scaler.go (hunk 25 above)
Comment (zpatrick) [reply api/provider/aws/environment_scaler.go]: why does this function exist?

File: api/provider/aws/test_aws/environment_scaler_test.go (hunk 27 above)
Comment (zpatrick) [reply api/provider/aws/test_aws/environment_scaler_test.go]: I totally disagree with @bivers. The old tests only testing the scaling algorithm (these need to be properly split up to achieve that), and had a ton of test cases. It's very difficult to get all of the old test cases to pass, because they were specifically designed to trip up a non-optimal algorithm. They should be added verbatim, or one needs to make a strong case as to why any should be left out.

File: common/models/scaler_run_info.go (hunk 35 above)
Comment (tlake) [reply common/models/scaler_run_info.go]: Ooh, I didn't catch this struct. I removed the usage of the struct in `environment_scaler.go` over in #440 PR, but not the struct. I'll fix that in #440.

File: common/models/resource.go (hunk 34 above)
Comment (tlake) [reply common/models/resource.go]: IIRC (and @nehayward please correct me if I'm wrong), but this was part of an experiment to see if it were absolutely necessary for `ResourceProvider` and `ResourceConsumer` to be two different things. I think the conclusion was that yes, they do need to be separate, and so this would want to go away.

File: api/provider/aws/environment_scaler.go (hunk 28 above)
Comment (tlake) [reply api/provider/aws/environment_scaler.go]: FWIW, I didn't see these sorts of functions ever defined as methods on a `ResourceProvider` or `ResourceConsumer`. I only ever saw these functions as standalone helper functions that took a provider and a consumer, and it was a little opaque in regards to what was happening without reading through the function itself.

The changes in #440 have refactored `HasResourcesFor` and `SubtractResourcesFor` into methods on `ResourceProvider`.

File: api/provider/aws/environment_scaler.go (hunk 26 above)
Comment (tlake) [reply api/provider/aws/environment_scaler.go]: This will be broken up into three functions as a result of #440.

File: api/provider/aws/environment_scaler.go (hunk 25 above)
Comment (tlake) [reply api/provider/aws/environment_scaler.go]: We don't know - this was something we found in `develop` and we weren't sure what the heck the point of it was.

I've removed it over in #440.

✅

File: api/provider/aws/environment_scaler.go (hunk 24 above)
Comment (tlake) [reply api/provider/aws/environment_scaler.go]: I think you'll find some relief over in #440. :D

File: api/provider/aws/environment_scaler.go (hunk 23 above)
Comment (tlake) [reply api/provider/aws/environment_scaler.go]: ✅ (#440)

File: common/models/resource.go (hunk 34 above)
Comment (tlake) [reply common/models/resource.go]: Now I understand what you all meant from a comment over in #440 - I'll make this change over there.

✅

Comment (nehayward) [reply common/models/resource.go]: Yup it was. @tlake.
